flask_session_captcha/__init__.py

```diff
@@ -1,97 +1,159 @@
+        self.include_alphabet = app.config.get("CAPTCHA_INCLUDE_ALPHABET", self.include_alphabet)
+        self.include_numeric = app.config.get("CAPTCHA_INCLUDE_NUMERIC", self.include_numeric)
+        self.include_punctuation = app.config.get("CAPTCHA_INCLUDE_PUNCTUATION", self.include_punctuation)
+
+
+        def _generate(*args, **kwargs) -> Markup:
+            """Generate Captcha Image"""
             if not self.enabled:
-                return ""
-            base64_captcha = self.generate()
-            data = "data:image/png;base64, {}".format(base64_captcha)
-            if css_class is not None:
-                return Markup("<img src='{}' class='{}'>".format(data, css_class))
-            return Markup("<img src='{}'>".format(data))
+                return Markup(" ")
+
+            base64_captcha = self.generate(*args, **kwargs)
+            data = f"data:image/png;base64, {base64_captcha}"
+            #TODO: replace it with argument base function: users can set ud,class,and ... in template in captcha:filter
+            css = f"class=\'{kwargs.get('css_class')}\'" if kwargs.get('css_class', None) else ''
+            return Markup(f"""<img src='{data}' {css} >""")
 
         app.jinja_env.globals['captcha'] = _generate
 
-        # Check for sessions that do not persist on the server. Issue a warning because
-        # they are most likely open to replay attacks. This addon is built upon flask-session.
-        session_type = app.config.get('SESSION_TYPE')
-        if session_type is None or session_type == "null":
-            raise RuntimeWarning(
-                "Flask-Sessionstore is not set to use a server persistent storage type. This likely means that captchas are vulnerable to replay attacks.")
-        elif session_type == "sqlalchemy":
-            # I have to do this as of version 0.3.1 of flask-session if using sqlalchemy as the session type in order to create the initial database.
-            # Flask-sessionstore seems to have the same problem.
-            app.session_interface.db.create_all()
-
-    def generate(self):
+    def generate(self,*args, **kwargs) -> str:
         """
-        Generates and returns a numeric captcha image in base64 format. 
-        Saves the correct answer in `session[self.session_key]`
-        Use later as:
+            generate captcha with given flags
+
+            :param:
+                numeric: if True generate captcha with numeric only
+                alphabet: if True generate captcha with alphabet
+                punctuation if True generate captcha with punctuation symbols
+
+            if both alphabet and numeric set to True this function generate captcha that contain both alphabet and numeric
 
-        src = captcha.generate()
-        <img src="{{src}}">
         """
-        answer = self.rand.randrange(self.max)
-        answer = str(answer).zfill(self.digits)
+
+        if not kwargs.get("numeric"):
```

Comment:
Haven't tested it yet, but this still seems to be an issue.

Reply:
``` python 



    def generate(self, *args, **kwargs) -> Markup:
        """Generate Captcha Image"""
        if not self.enabled:
            return Markup(" ")

        # if none of options passed by developer set all options to False
        base64_captcha = self.__generate(include_alphabet=kwargs.get(__key="include_alphabet", __default=False),
                                         include_punctuation=kwargs.get(__key="include_punctuation", __default=False),
                                         include_numeric=kwargs.get(__key="include_numeric", __default=False))

        data = f"data:image/png;base64, {base64_captcha}"
        css = f"class=\'{kwargs.get('css_class')}\'" if kwargs.get('css_class', None) else ''
        return Markup(f"<img src='{data}' {css} >")

    def __generate(self, include_alphabet: bool, include_punctuation: bool, include_numeric: bool) -> str:
        """
        generate captcha with given flags
        Don't call this method Directly. use self.generate() instead

            :param:
                include_numeric: if True generate captcha with numeric only
                include_alphabet: if True generate captcha with alphabet
                include_punctuation if True generate captcha with punctuation symbols

            :returns:str:
                string value of captcha encoded in base64

       this method generate captcha with default options:
                Baseconfig.include_alphabet: bool = True
                Baseconfig.include_numeric: bool = True
                Baseconfig.include_punctuation: bool = False


        if you want to generate a single captcha with different option you can change it with passing these arguments to
        template filter

        example:
            {{  captcha() }} # generate a captcha with default config or options that sets in app.config
            {{  captcha(include_numeric=True) }} # generate a captcha with only #numeric
            {{  captcha(include_alphabet=True) }} # generate a captcha with only #alphabet

        """
        answer = []
        args = [include_numeric, include_alphabet, include_punctuation]
        args = [any([each]) for each in args] # make sure all inputs are valid <True, False>
        is_arg_passed = sum(args)

        if not is_arg_passed:  # if options not passed by developer use default config
            include_numeric = self.include_numeric
            include_alphabet = self.include_alphabet
            include_punctuation = self.include_punctuation

        # single mode captcha options
        if include_numeric and not include_alphabet and not include_punctuation:  # only numeric captcha
            answer += self.random_number(length=self.length)
        if include_alphabet and not include_numeric and not include_punctuation:  # only alphabetical captcha
            answer += self.random_alphabet(length=self.length)
        if include_punctuation and not include_numeric and not include_alphabet:  # only punctuation captcha
            answer += self.random_punctuation(length=self.length)

        if len(answer) == 0:  # mix captcha mode <developer passed more than one option>
            selected = {}
            # determine which options passed by developer
            if include_punctuation:
                selected["punctuation"] = self.random_punctuation
            if include_numeric:
                selected["numeric"] = self.random_number
            if include_alphabet:
                selected["alphabet"] = self.random_alphabet

            each_option_char_len = self.length // is_arg_passed
            # calculate evenly space for each option base on options

            for each in selected:
                answer += (selected[each](length=each_option_char_len))

            answer += (self.random_alphabet(
                length=(self.length - len(answer))))  # fill gap with alphabet if is_arg_passed wsa odd number
            self.shuffle_list(answer)

        answer = "".join(answer)
        image_data = self.image_generator.generate(answer)
        base64_captcha = base64.b64encode(image_data.getvalue()).decode("ascii")
        self.set_in_session(key=self.session_key, value=answer)
        self.debug_log(f"Captcha Generated. key: {answer}")
        return base64_captcha

```
how about this ?
In this case, the entries are validated and if the entry is not given, the basic settings are used, otherwise, the entry is preferred to the basic settings.

Reply:
I recommend you add a test for this behaviour. I tested the previous version manually and it crashed on `{{ captcha(numeric=True) }}` in the template.

The description/naming convention of the arguments is a bit confusing. `include` does not mean only.
```python
#include_numeric: if True generate captcha with numeric only 
#include_alphabet: if True generate captcha with alphabet 
#include_punctuation if True generate captcha with punctuation symbols
```
Otherwise sure, that could work.